hbase/src/main/scala/akka/stream/alpakka/hbase/internal/HBaseFlowStage.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.hbase.internal
+
+import akka.stream._
+import akka.stream.alpakka.hbase.HTableSettings
+import akka.stream.stage._
+import org.apache.hadoop.hbase.client.Table
+
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success, Try }
+
+private[hbase] class HBaseFlowStage[A](settings: HTableSettings[A])
+    extends GraphStage[FlowShape[A, A]]
+    with HBaseCapabilities {
+
+  override protected def initialAttributes: Attributes =
+    Attributes.name("HBaseFLow").and(ActorAttributes.dispatcher("akka.stream.default-blocking-io-dispatcher"))
+
+  private val in = Inlet[A]("messages")
+  private val out = Outlet[A]("result")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with StageLogging {
+
+      implicit val connection = connect(settings.conf)
+
+      val table: Table = getOrCreateTable(settings.tableName, settings.columnFamilies)
+
+      var completionState: Option[Try[Unit]] = None
+
+      setHandler(out, new OutHandler {
+        override def onPull() =
+          tryPull(in)
+      })
+
+      setHandler(in, new InHandler {
+        override def onPush() = {
+          val msg = grab(in)
+
+          table.put(settings.converter(msg))
+
+          push(out, msg)
+        }
+
+        override def onUpstreamFinish() = {
+          completionState = Some(Success(()))
+          checkForCompletion()
+        }
+
+        override def onUpstreamFailure(ex: Throwable) = {
+          completionState = Some(Failure(ex))
+          checkForCompletion()
+        }
+
+      })
+
+      def checkForCompletion() =
```

Comment:
The original behaviour of GraphStageLogic is to complete `onUpstreamFinish` and to fail `onUpstreamFailure` so this is now just a more convoluted way of doing the same, I think you can remove this method, the `completionState` and the overrides.